distributed/s3.py

```diff
@@ -26,148 +32,635 @@
-    path_delimiter: string (optional)
-        Delimiter like ``'/'`` to define implicit S3 directory structure
+    fn: string
+        location in S3
     executor: Executor (optional)
         defaults to most recently created executor
+    s3: HDFileSystem (optional)
     lazy: boolean (optional)
         If True then return lazily evaluated dask Values
-    anon: boolean (optional)
-        If True then don't try to authenticate with AWS
+    delimiter: bytes
+        An optional delimiter, like ``b'\n'`` on which to split blocks of bytes
+    not_zero: force seek of start-of-file delimiter, discarding header
+    blocksize: int (=128MB)
+        Chunk size
+    **s3pars: keyword arguments
+        Extra keywords to send to boto3 session (anon, key, secret...)
 
     Returns
     -------
-    list of Futures.  Each future holds bytes for one key within the bucket
-
-    Examples
-    --------
-    >>> futures = read_bytes('distributed-test', 'test')  # doctest: +SKIP
-    >>> futures  # doctest: +SKIP
-    [<Future: status: finished, key: read_content_from_keys-00092e8a75141837c1e9b717b289f9d2>,
-     <Future: status: finished, key: read_content_from_keys-4f0f2cbcf4573a373cc62467ffbfd30d>]
-    >>> futures[0].result()  # doctest: +SKIP
-    b'{"amount": 100, "name": "Alice"}\\n{"amount": 200, "name": "Bob"}\\n
-      {"amount": 300, "name": "Charlie"}\\n{"amount": 400, "name": "Dennis"}\\n'
+    List of ``distributed.Future`` objects if ``lazy=False``
+    or ``dask.Value`` objects if ``lazy=True``
     """
+    if s3 is None:
+        s3 = S3FileSystem(**s3pars)
     executor = default_executor(executor)
-    s3_objects = get_list_of_summary_objects(bucket_name, prefix,
-                                             path_delimiter, anon=anon)
-    keys = [obj.key for obj in s3_objects]
-
-    names = ['read-bytes-{0}'.format(key) for key in keys]
-
+    filenames, lengths, offsets = [], [], []
+    for afile in s3.glob(fn):
+        size = s3.info(afile)['Size']
+        offset = list(range(0, size, blocksize))
+        if not_zero:
+            offset[0] = 1
+        offsets.extend(offset)
+        filenames.extend([afile]*len(offset))
+        lengths.extend([blocksize]*len(offset))
+    names = ['read-binary-s3-%s-%s' % (fn, tokenize(offset, length, delimiter, not_zero))
+            for fn, offset, length in zip(filenames, offsets, lengths)]
+
+    logger.debug("Read %d blocks of binary bytes from %s", len(names), fn)
     if lazy:
-        values = [Value(name, [{name: (read_content_from_keys, bucket_name,
-                                       key, anon)}])
-                  for name, key in zip(names, keys)]
+        executor._send_to_scheduler({'op': 'update-graph',
+                                     'tasks': {},
+                                     'dependencies': set(),
+                                     'keys': [],
+                                     'client': executor.id})
```

Comment:
This send_to_scheduler call is not necessary.  It was used in hdfs to communicate restrictions, which don't exist here.